From bc72eeb2be0b1a17c7591dfa0548c326d8d59fa0 Mon Sep 17 00:00:00 2001
From: Vignesh R <vigneshr@ti.com>
Date: Fri, 30 Sep 2016 11:05:33 +0530
Subject: [PATCH 3/3] serial: 8250: omap: Enable UART module wakeup based on
 device_may_wakeup() status

Enable/Clear module level UART wakeup in UART_OMAP_WER register based on
return value of device_may_wakeup() in .suspend(). This is allows
userspace to use sysfs to control the ability of UART to wakeup the
system from low power state. Register is restored back in .startup()
call that happens as part of resume sequence.

With this patch, userspace can control UART wakeup capability via sysfs:
To enable wakeup capability:
echo enabled >  /sys/class/tty/ttyXX/device/power/wakeup
For disabling wakeup capability:
echo disabled > /sys/class/tty/ttyXX/device/power/wakeup

Signed-off-by: Vignesh R <vigneshr@ti.com>
---
 drivers/tty/serial/8250/8250_omap.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/tty/serial/8250/8250_omap.c b/drivers/tty/serial/8250/8250_omap.c
index 6f76051..a47273e 100644
--- a/drivers/tty/serial/8250/8250_omap.c
+++ b/drivers/tty/serial/8250/8250_omap.c
@@ -1303,8 +1303,18 @@ static void omap8250_complete(struct device *dev)
 static int omap8250_suspend(struct device *dev)
 {
 	struct omap8250_priv *priv = dev_get_drvdata(dev);
+	struct uart_8250_port *up = serial8250_get_port(priv->line);
 
 	serial8250_suspend_port(priv->line);
+
+	pm_runtime_get_sync(dev);
+	if (device_may_wakeup(dev))
+		serial_out(up, UART_OMAP_WER, priv->wer);
+	else
+		serial_out(up, UART_OMAP_WER, 0);
+	pm_runtime_mark_last_busy(dev);
+	pm_runtime_put_autosuspend(dev);
+
 	flush_work(&priv->qos_work);
 	return 0;
 }
@@ -1384,6 +1394,8 @@ static int omap8250_runtime_suspend(struct device *dev)
 	}
 
 	if (priv->habit & UART_ERRATA_CLOCK_DISABLE) {
+		/* Save module level wakeup register */
+		u32 wer = serial_in(up, UART_OMAP_WER);
 		int ret;
 
 		ret = omap8250_soft_reset(dev);
@@ -1392,6 +1404,8 @@ static int omap8250_runtime_suspend(struct device *dev)
 
 		/* Restore to UART mode after reset (for wakeup) */
 		omap8250_update_mdr1(up, priv);
+		/* Restore module level wakeup register */
+		serial_out(up, UART_OMAP_WER, wer);
 	}
 
 	if (up->dma && up->dma->rxchan)
-- 
2.9.3

